datafusion/core/src/physical_optimizer/remove_unnecessary_sorts.rs

```diff
@@ -0,0 +1,887 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! RemoveUnnecessarySorts optimizer rule inspects SortExec's in the given
+//! physical plan and removes the ones it can prove unnecessary. The rule can
+//! work on valid *and* invalid physical plans with respect to sorting
+//! requirements, but always produces a valid physical plan in this sense.
+//!
+//! A non-realistic but easy to follow example: Assume that we somehow get the fragment
+//! "SortExec: [nullable_col@0 ASC]",
+//! "  SortExec: [non_nullable_col@1 ASC]",
+//! in the physical plan. The first sort is unnecessary since its result is overwritten
+//! by another SortExec. Therefore, this rule removes it from the physical plan.
+use crate::error::Result;
+use crate::physical_optimizer::utils::{
+    add_sort_above_child, ordering_satisfy, ordering_satisfy_concrete,
+};
+use crate::physical_optimizer::PhysicalOptimizerRule;
+use crate::physical_plan::rewrite::TreeNodeRewritable;
+use crate::physical_plan::sorts::sort::SortExec;
+use crate::physical_plan::windows::WindowAggExec;
+use crate::physical_plan::{with_new_children_if_necessary, ExecutionPlan};
+use crate::prelude::SessionConfig;
+use arrow::datatypes::SchemaRef;
+use datafusion_common::{reverse_sort_options, DataFusionError};
+use datafusion_physical_expr::{PhysicalExpr, PhysicalSortExpr};
+use itertools::izip;
+use std::iter::zip;
+use std::sync::Arc;
+
+/// This rule inspects SortExec's in the given physical plan and removes the
+/// ones it can prove unnecessary.
+#[derive(Default)]
+pub struct RemoveUnnecessarySorts {}
+
+impl RemoveUnnecessarySorts {
+    #[allow(missing_docs)]
+    pub fn new() -> Self {
+        Self {}
+    }
+}
+
+/// This is a "data class" we use within the [RemoveUnnecessarySorts] rule
+/// that tracks the closest `SortExec` descendant for every child of a plan.
```

Comment:
I wonder if we could use `ExecutionPlan::output_ordering()` rather than tracking the state separately?

That way the logic of which nodes retain what ordering is kept within the execution plan nodes themselves rather than in this pass.

Though it looks strange to me to have both `ExecutionPlan::maintains_input_order` as well as `ExecutionPlan::output_order` (I think the maintains_input_order came before) -- maybe we should remove `ExecutionPlan::maintains_input_order` 🤔  -- I could try this if you think it worthwhile

Reply:
The reason we use this struct is to be able to remove `SortExec`s down below the plan. By using `output_ordering` we can remove unnecessary `SortExec`, from the kind of plans where sorting and its usage are immediate like `SortExec` -> `WindowAggExec`. However, to remove unnecessary `SortExec` where sorting and its usage are linked with other order preserving executor(s) like `SortExec` -> `ORDER_PRESERVING_EXEC` -> `WindowAggExec` we need to keep track of linkage from current executor till to the corresponding `SortExec`. In the current implementation, Datafusion may be is not generating, these kind of plans. However, we didn't want to overfit to a strong assumption. As you say may be `maintains_input_order` is now unnecessary and this information can be derived by comparing `output_ordering` of consecutive executors. However, I think it is still helpful to have as a convenience method.

Reply:
The linkage should be tracked by the `output_ordering()` method also. If an operator is order preserving, it should override the method `output_ordering()` and `maintains_input_order()` and make it return `self.input.output_ordering()`